### DISCUSSION TITLE
Examples of how to register a reply function

### DISCUSSION THREAD

**Discussion by sonichi** [microsoft/autogen#1517] [General]
@xmi1an you can hook in function that is called whenever an agent receives a message. 

Assuming you have two agents `user_proxy` and `assistant`, you can [register a function](https://microsoft.github.io/autogen/docs/reference/agentchat/conversable_agent#conversableagent-objects) with these agents that gets called with some parameters. 

```python

def print_messages(recipient, messages, sender, config): 
    if "callback" in config and  config["callback"] is not None:
        callback = config["callback"]
        callback(sender, recipient, messages[-1])
    print(f"Messages sent to: {recipient.name} | num messages: {len(messages)}")
    return False, None  # required to ensure the agent communication flow continues

user_proxy.register_reply(
    [autogen.Agent, None],
    reply_func=print_messages, 
    config={"callback": None},
)

assistant.register_reply(
    [autogen.Agent, None],
    reply_func=print_messages, 
    config={"callback": None},
) 
``` 

In the above, we register a `print_messages` message that is called each time the agent receive a message.

_Originally posted by @victordibia in https://github.com/microsoft/autogen/issues/478#issuecomment-1785922097_

It's a good example to add to FAQ.

**Comment by abdullamatar** [comment]
Hi, can you please elaborate as to why returning `False, None` is required to continue the conversation flow? From the `register_reply()` Args documentation the following line is present under the `trigger` argument:
>  \- If a callable is provided, the reply function will be called when the callable returns True. 

However as noted above the `reply_func` must return `False, None` and the `trigger` in this case is the `autogen.Agent` class.

**Comment by ekzhu** [comment]
> Hi, can you please elaborate as to why returning `False, None` is required to continue the conversation flow? From the `register_reply()` Args documentation the following line is present under the `trigger` argument:
> 
> > - If a callable is provided, the reply function will be called when the callable returns True.
> 
> However as noted above the `reply_func` must return `False, None` and the `trigger` in this case is the `autogen.Agent` class.

The triggers decides whether to enter the registered reply function. In this case the trigger basically says any message from any agent is accepted and will enter the reply function.

Autogen uses a pipeline of register reply functions to handle incoming messages. If one reply function gets triggered, it will process the message, and returns a boolean signal. If the signal is True, then the rest of the pipeline gets short-circuited and the message generated from the triggered reply funciton will be returned. If the signal is False, then the message continue to the next reply function in the pipeline. 

In this example, the message gets passed on to the next reply function, because the signal is False. However the message was processed (i.e., printed). It is just that the control gets handed back to the pipeline.

**Comment by abdullamatar** [comment]
Thank you :)

**Comment by che-shr-cat** [comment]
Hi everyone! I have a couple of questions regarding the reply functions.

1. A reply_function gets a list of messages, among other parameters. Is it safe to assume that if we are interested in the freshest message only (say, to draw it in the interface),  we can take messages[-1], and we do not miss anything? There are no cases when more than one message comes in a single call. Is that the case?

2. The reply_function will not receive the last message in the group chat, as after that message is generated, no one will receive it, and no reply_function will be triggered as a result. Is it true or am I missing anything?

A related question: Are there any plans to add a callback after generating each message? Or maybe there are other ways on how to get access to the latest messages as soon as they are generated?
